src/host/stream.cpp

```diff
@@ -115,64 +106,54 @@ using Microsoft::Console::Interactivity::ServiceLocator;
+                    (WI_IsFlagSet(pInputBuffer->InputMode, ENABLE_VIRTUAL_TERMINAL_INPUT) ||
+                     (Event.Event.KeyEvent.wVirtualKeyCode != VK_ESCAPE &&
+                      Event.Event.KeyEvent.uChar.UnicodeChar != UNICODE_LINEFEED)))
                 {
-                    *pwchOut = keyEvent->GetCharData();
+                    *pwchOut = Event.Event.KeyEvent.uChar.UnicodeChar;
                     return STATUS_SUCCESS;
                 }
             }
 
-            if (keyEvent->IsKeyDown())
+            if (Event.Event.KeyEvent.bKeyDown)
             {
                 if (pCommandLineEditingKeys && commandLineEditKey)
                 {
                     *pCommandLineEditingKeys = true;
-                    *pwchOut = static_cast<wchar_t>(keyEvent->GetVirtualKeyCode());
+                    *pwchOut = static_cast<wchar_t>(Event.Event.KeyEvent.wVirtualKeyCode);
                     return STATUS_SUCCESS;
                 }
-                else if (pPopupKeys && commandLineEditKey)
+
+                if (pPopupKeys && commandLineEditKey)
                 {
                     *pPopupKeys = true;
-                    *pwchOut = static_cast<char>(keyEvent->GetVirtualKeyCode());
+                    *pwchOut = static_cast<char>(Event.Event.KeyEvent.wVirtualKeyCode);
                     return STATUS_SUCCESS;
                 }
-                else
-                {
-                    const auto zeroVkeyData = OneCoreSafeVkKeyScanW(0);
-                    const auto zeroVKey = LOBYTE(zeroVkeyData);
-                    const auto zeroControlKeyState = HIBYTE(zeroVkeyData);
 
-                    try
-                    {
-                        // Convert real Windows NT modifier bit into bizarre Console bits
-                        auto consoleModKeyState = FromVkKeyScan(zeroControlKeyState);
+                const auto zeroKey = OneCoreSafeVkKeyScanW(0);
 
-                        if (zeroVKey == keyEvent->GetVirtualKeyCode() &&
-                            keyEvent->DoActiveModifierKeysMatch(consoleModKeyState))
-                        {
-                            // This really is the character 0x0000
-                            *pwchOut = keyEvent->GetCharData();
-                            return STATUS_SUCCESS;
-                        }
-                    }
-                    catch (...)
-                    {
-                        LOG_HR(wil::ResultFromCaughtException());
-                    }
+                if (LOBYTE(zeroKey) == Event.Event.KeyEvent.wVirtualKeyCode &&
+                    WI_IsAnyFlagSet(Event.Event.KeyEvent.dwControlKeyState, ALT_PRESSED) == WI_IsFlagSet(zeroKey, 0x400) &&
+                    WI_IsAnyFlagSet(Event.Event.KeyEvent.dwControlKeyState, CTRL_PRESSED) == WI_IsFlagSet(zeroKey, 0x200) &&
+                    WI_IsAnyFlagSet(Event.Event.KeyEvent.dwControlKeyState, SHIFT_PRESSED) == WI_IsFlagSet(zeroKey, 0x100))

```

Comment:
This is what replaces the `std::unordered_set` code. I've decided against using clever bitmasking or against converting `dwControlKeyState` into a Win32 compatible bitset first, because I felt like writing it this way is easier to understand (and review). The constants at the end are kind of magic, but it should be easy to see how they correspond to the ALT_PRESSED, CTRL_PRESSED, SHIFT_PRESSED constants.